File: FinalProject.py
#!/usr/bin/env python
# coding: utf-8

# In[2]:


import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Read the dataset from a csv file
df = pd.read_csv("My_Data_Set.csv", encoding='ISO-8859-1')

# Print the structure of the dataset
print(df.info())


# In[3]:


# decribe the summary of the dataset
print(df.describe(include='all'))


# In[4]:


# Calculate the middle index
middle_index = len(df) // 2
print("No of total rows" , len(df))
print()

# Display the middle 10 rows of the dataset
print(df.iloc[middle_index - 5 : middle_index + 5])


# In[5]:


print("Before modifications:", df.shape, "  ", end ="")
# drop rows with empty cells or "."
df = df.dropna(subset=['Risk Adjuested Mortality Rate', 
                       '# of Deaths', '# of Cases','Hospital Ratings'], how='any')
df = df[~df.isin(['.']).any(axis=1)]
print("After modifications:", df.shape)
print()

# Display the middle 10 rows of the dataset
# Display the middle 10 rows of the dataset
print(df.iloc[middle_index - 5 : middle_index + 5])


# In[6]:


# Print the structure of the dataset
print(df.info())

Though the Risk Adjuested Mortality Rate , # of Deaths, # of Cases, LONGITUDE, LATITUDE variables are numeric but they are considered as object. So change the data type for further analysis 
# In[7]:


# Convert required columns to numeric
df['Risk Adjuested Mortality Rate'] = pd.to_numeric(
    df['Risk Adjuested Mortality Rate'], errors='coerce')
df['# of Deaths'] = pd.to_numeric(df['# of Deaths'], errors='coerce')
df['# of Cases'] = pd.to_numeric(df['# of Cases'], errors='coerce')
df['LONGITUDE'] = pd.to_numeric(df['LONGITUDE'], errors='coerce')
df['LATITUDE'] = pd.to_numeric(df['LATITUDE'], errors='coerce')


# In[8]:


# Save the modified dataset as a new CSV file
df.to_csv('modified_dataset.csv', index=False)
print("Saved the modifications")


# In[9]:


# Print the column names
print(df.columns)

# Replace spaces with underscores in column names
df.columns = df.columns.str.replace(" ", "_")

# Print the modified column names
print(df.columns)


# In[10]:


# Histogram of quality ratings
sns.histplot(df['Hospital_Ratings'], bins=10)
plt.xlabel('Hospital_Ratings')
plt.ylabel('Count')
plt.title('Histogram of Quality Ratings')
plt.show()


# In[11]:


# Histogram of risk-adjusted mortality rates
sns.histplot(df['Risk_Adjuested_Mortality_Rate'], bins=10)
plt.xlabel('Risk_Adjuested_Mortality_Rate')
plt.ylabel('Count')
plt.title('Histogram of Risk-Adjusted Mortality Rates')
plt.show()


# In[12]:


# plot of quality ratings by procedure/condition
sns.histplot(df['Procedure/Condition'], bins=10)
plt.xlabel('Procedure/Condition')
plt.xticks(rotation=90)
plt.ylabel('Count')
plt.title('Histogram of Risk-Adjusted Mortality Rates')
plt.show()


# In[16]:


# Filter the relevant columns
data = df[["HOSPITAL", "Procedure/Condition", 
           "Risk_Adjuested_Mortality_Rate", "Hospital_Ratings"]]

# Group the data by hospital and procedure/condition, 
# and calculate the mean mortality rate for each group
grouped_data = data.groupby(
    ["HOSPITAL", "Procedure/Condition"]).mean().reset_index()

# Join the grouped data with the hospital ratings
joined_data = pd.merge(grouped_data, data[
    ["HOSPITAL", "Hospital_Ratings"]].drop_duplicates(), on="HOSPITAL")

# Filter out rows with null or "None" values in the "Hospital_Ratings" column
joined_data = joined_data[(joined_data["Hospital_Ratings"].notnull()) &
                          (joined_data["Hospital_Ratings"] != "None")]

# Calculate the mean mortality rate for hospitals with each rating
mean_rates = joined_data.groupby(
    "Hospital_Ratings")["Risk_Adjuested_Mortality_Rate"].mean()

# Print the results
print(mean_rates)


# In[19]:


# Filter the data to include only relevant columns and rows
filtered_data = data[['Procedure/Condition', 
                     'Risk_Adjuested_Mortality_Rate', 'Hospital_Ratings']]

# Exclude rows where Hospital_Ratings is None
filtered_data = filtered_data[filtered_data['Hospital_Ratings'] != 'None']

# Group the data by hospital ratings 
# and calculate mean risk-adjusted mortality rate for each procedure/condition
grouped_data = filtered_data.groupby(
   ['Hospital_Ratings', 'Procedure/Condition']).mean()

# Plot the results
fig, ax = plt.subplots(figsize=(12, 8))
grouped_data.unstack().plot(ax=ax, kind='bar')
plt.legend(title='Procedure/Condition')
plt.ylabel('Mean Risk-Adjusted Mortality Rate')
plt.xlabel('Hospital Ratings')
plt.title('Relationship b/w Hospital Ratings and Risk-Adjusted Mortality Rates')
plt.show()


# In[226]:


data= df

# Extract unique cells from a column
procedures = data["Procedure/Condition"].unique() # List of procedures/conditions
hospitals = data[data["Procedure/Condition"].isin(procedures)]


# Filter the hospitals based on their quality ratings
quality_ratings = ["Better", "As Expected", "Worse"] # List of quality ratings
hospitals = hospitals[hospitals["Hospital_Ratings"].isin(quality_ratings)]

# Calculate the mean risk-adjusted mortality rates for each combination of procedure/condition and quality rating
result = hospitals.groupby(["Procedure/Condition", "Hospital_Ratings"])["Risk_Adjuested_Mortality_Rate"].mean()

# Pivot the data to get mean risk-adjusted mortality rates for each combination of procedure/condition and quality rating
pivot_table = pd.pivot_table(hospitals, values="Risk_Adjuested_Mortality_Rate", index="Procedure/Condition", columns="Hospital_Ratings", aggfunc=np.mean)

print(pivot_table)

Based on the output of the code, we can see that hospitals with higher quality ratings are generally associated with lower risk-adjusted mortality rates for specific procedures/conditions. This can be observed from the fact that the "As Expected" and "Better" mortality rates are generally lower for hospitals with higher quality ratings, while the "Worse" mortality rates are generally higher for hospitals with lower quality ratings. However, this is not always the case, as there are some exceptions where hospitals with higher quality ratings have higher mortality rates for certain procedures/conditions, such as "Acute Stroke Subarachnoid". Overall, we can say that there is a correlation between hospital quality ratings and risk-adjusted mortality rates for specific procedures/conditions
# In[ ]:

File: FinalProject.R
---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(readr)
library(ggplot2)

```

```{r}
# Read in the dataset
california_hospitals <- read_csv("/Users/dhavani/Downloads/modified_dataset.csv")
str(california_hospitals)
```

```{r}
# Summary statistics for modified_dataset
summary(california_hospitals)
```

```{r}
# view the original column names
colnames(california_hospitals)
```

```{r}
# replace spaces with underscores in column names
colnames(california_hospitals) <- gsub(" ", "_", colnames(california_hospitals))
# replace spaces with underscores in column names
colnames(california_hospitals) <- gsub("/", "Or", colnames(california_hospitals))
# replace spaces with underscores in column names
colnames(california_hospitals) <- gsub("#", "No", colnames(california_hospitals))
```

```{r}
# view the updated column names
colnames(california_hospitals)
```
```{r message=FALSE, warning=FALSE}
# Create visualizations for YEAR
ggplot(california_hospitals, aes(x = YEAR)) +
  geom_histogram() +
  labs(x = "Year", y = "Count", title = "Distribution of Year")

```
```{r}
ggplot(california_hospitals, aes(x = COUNTY)) + 
  geom_bar(fill = "red") + 
  labs(x = "County", y = "Count", title = "Distribution of County") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```
```{r}
ggplot(california_hospitals, aes(x = ProcedureOrCondition)) +
  geom_bar(fill = "9999FF") +
  labs(x = "Procedure/Condition", y = "Count", title = "Distribution of Procedure/Condition")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```
```{r message=FALSE, warning=FALSE}
ggplot(california_hospitals, aes( x = Risk_Adjuested_Mortality_Rate)) +
  geom_histogram(fill = "blue") +
  labs(x ="Risk_Adjuested_Mortality_Rate", title ="Distribution of Risk_Adjuested_Mortality_Rate")
```

```{r}
library(cowplot)
# plot for Risk_Adjuested_Mortality_Rate <= 50
plot1 <- ggplot(california_hospitals[!is.na(california_hospitals$Risk_Adjuested_Mortality_Rate) & california_hospitals$Risk_Adjuested_Mortality_Rate <= 50,], 
                aes(x = Risk_Adjuested_Mortality_Rate)) +
  geom_histogram(binwidth = 1, fill = "blue", alpha = 0.5) +
  scale_x_continuous(breaks = seq(0, 50, 5)) +
  labs(x = "Morality Rate (<= 50)", y = "Count", title = "Distribution of Morality Rate") +
  theme_bw()

# plot for Risk_Adjuested_Mortality_Rate > 50
plot2 <- ggplot(california_hospitals[!is.na(california_hospitals$Risk_Adjuested_Mortality_Rate) & california_hospitals$Risk_Adjuested_Mortality_Rate > 50,], 
                aes(x = Risk_Adjuested_Mortality_Rate)) +
  geom_histogram(binwidth = 1, fill = "blue", alpha = 0.5) +
  scale_x_continuous(breaks = seq(50, 100, 5)) +
  labs(x = "Morality Rate (> 50)", y = "Count", title = "Distribution of Morality Rate") +
  theme_bw()


plot_grid(plot1, plot2, ncol = 2, align = "h")
```

Question 2

```{r}
# Read in the dataset
hospital_data <- california_hospitals
ca_hospitals <- subset(hospital_data, COUNTY != "AAAA")

# Calculate the average risk-adjusted mortality rate for each procedure/condition
ca_avg_mortality <- aggregate(Risk_Adjuested_Mortality_Rate ~ ProcedureOrCondition,
                              data = ca_hospitals, FUN = mean)

ca_avg_mortality

```
We can create a bar plot to show the average risk-adjusted mortality rate for each procedure/condition. This will give us an overview of which procedures/conditions tend to have higher or lower mortality rates overall.

```{r}

#Bar plot of average mortality rates by procedure/condition
ggplot(ca_avg_mortality, aes(x = ProcedureOrCondition, 
                             y = Risk_Adjuested_Mortality_Rate)) +
  geom_bar(stat = "identity", fill = "orange") +
  xlab("Procedure/Condition") +
  ylab("Average Risk-Adjusted Mortality Rate") +
  ggtitle("Average Mortality Rates by Procedure/Condition in California Hospitals")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r}
# Merge the average mortality rates with the hospital data
ca_hospitals <- merge(ca_hospitals, ca_avg_mortality, 
                      by = "ProcedureOrCondition", suffixes = c("", ".statewide"))

```

Scatter plot of hospital mortality rates vs. statewide average mortality rates
We can create a scatter plot to compare each hospital's risk-adjusted mortality rate for a given procedure/condition with the statewide average for that procedure/condition. This will allow us to see which hospitals have higher or lower mortality rates than the statewide average, and how far away they are from the average.


```{r}
ggplot(ca_hospitals, aes(x = Risk_Adjuested_Mortality_Rate.statewide, y = Risk_Adjuested_Mortality_Rate)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "red") +
  xlab("Statewide Average Risk-Adjusted Mortality Rate") +
  ylab("Hospital Risk-Adjusted Mortality Rate") +
  ggtitle("Hospital Mortality Rates vs. Statewide Average by Procedure/Condition in California") +
  facet_wrap(~ ProcedureOrCondition, ncol = 3, scales = "free")

```


```{r}

# Create a new column to indicate if the hospital has a higher 
#or lower mortality rate than the statewide average
ca_hospitals$Mortality.Rate.Comparison <- 
  ifelse(
    ca_hospitals$Risk_Adjuested_Mortality_Rate > 
      ca_hospitals$Risk_Adjuested_Mortality_Rate.statewide, "Higher", ifelse(ca_hospitals$Risk_Adjuested_Mortality_Rate <                                                                ca_hospitals$Risk_Adjuested_Mortality_Rate.statewide, "Lower", "Average"))

# Print the hospitals that have consistently higher or
#lower mortality rates than the statewide average for a particular procedure/condition
consistent_hospitals <- subset(ca_hospitals, 
                               Mortality.Rate.Comparison != "Average")


```

```{r}
ggplot(ca_hospitals, aes(x = ProcedureOrCondition, 
                         fill = Mortality.Rate.Comparison)) +
  geom_bar() +
  xlab("Procedure/Condition") +
  ylab("Number of Hospitals") +
  ggtitle(
    "Distribution of Hospitals by Mortality Rate Comparison in California") +
  facet_wrap(~ Mortality.Rate.Comparison, ncol = 1, scales = "free_y") +
  scale_fill_manual(values = c("Higher" = "darkred", 
                               "Lower" = "seagreen", "Average" = "gray"))+  theme(axis.text.x = element_text(angle = 90, hjust = 1))


```
These plots provide a visual summary of the results we obtained earlier. We can see that some procedures/conditions, tend to have higher mortality rates overall, while others, tend to have lower mortality rates. We can also see that there are some hospitals that consistently have higher or lower mortality rates than the statewide average for a given procedure/condition, while others are closer to the average.

Question 3

To analyze the relationship between the geographical location of a hospital and its risk-adjusted mortality rate for specific procedures/conditions, we can start by creating a scatterplot.

First import required libraries Next, we can filter the data to only include the necessary columns for our analysis, and remove any rows with missing latitude or longitude data:


```{r}
# Load necessary libraries
library(ggplot2)
library(dplyr)
library(ggplot2)
library(tidyr)
library(lmtest)

hospital_data <- california_hospitals

relevant_data <- hospital_data[, c("COUNTY", "HOSPITAL",
                                   "ProcedureOrCondition",
                                   "Risk_Adjuested_Mortality_Rate",
                                   "LONGITUDE", "LATITUDE")]

relevant_data <- relevant_data[!is.na(relevant_data$LONGITUDE)
                               & !is.na(relevant_data$LATITUDE), ]

```

Now, we can create a scatterplot using ggplot2, with longitude and latitude on the x and y axes, respectively, and risk-adjusted mortality rate represented by color:

```{r}
ggplot(data = relevant_data, aes(x = LONGITUDE, y = LATITUDE,
                                 color = Risk_Adjuested_Mortality_Rate)) + 
  geom_point() +
  labs(title = "Relationship between Hospital Location 
       and Risk-Adjusted Mortality Rate", 
       x = "Longitude", y = "Latitude",
       color = "Risk-Adjusted Mortality Rate")

```
The resulting plot will show whether there is any relationship between hospital location and risk-adjusted mortality rate, with higher mortality rates represented by darker colors.

In addition to the scatterplot, we can also calculate the correlation coefficient between latitude/longitude and risk-adjusted mortality rate using the cor() function

This will give us a numeric value for the correlation, with positive values indicating a positive correlation (i.e., higher mortality rates in areas with higher latitude/longitude values) and negative values indicating a negative correlation (i.e., higher mortality rates in areas with lower latitude/longitude values).

```{r}
cor(relevant_data[c("LONGITUDE", "LATITUDE")], 
    relevant_data$Risk_Adjuested_Mortality_Rate)
```

Based on the correlation output you provided, it appears that there is a weak correlation between the geographical location of a hospital and its risk-adjusted mortality rate for specific procedures/conditions. The correlation coefficient for longitude and risk-adjusted mortality rate is negative, indicating a very weak negative relationship, while the correlation coefficient for latitude and risk-adjusted mortality rate is positive, indicating a very weak positive relationship.

Now lets build a regression model for further analysis 

```{r}

# Filter for specific procedure/condition
my_procedure <- "AMI"
# You can replace this with any other procedure/condition
filtered_data <- hospital_data %>%
  filter(`ProcedureOrCondition` == my_procedure)

# Remove rows with missing location data
filtered_data <- filtered_data %>%
  drop_na(LONGITUDE, LATITUDE)

# Run linear regression model
model <- lm(Risk_Adjuested_Mortality_Rate ~ LONGITUDE + LATITUDE,
            data = filtered_data)

# Check model assumptions
plot(model)

```

```{r}
# Check model significance
summary(model)
```

Based on the output of the regression model, we can conclude that there is a statistically significant relationship between the geographical location of a hospital (latitude and longitude) and its risk-adjusted mortality rate for specific procedures/conditions.

The coefficients for both longitude and latitude are positive, indicating that as the longitude and latitude of a hospital increase, its risk-adjusted mortality rate also tends to increase.

However, it is important to note that the R-squared value for the model is very low (0.005545), meaning that only a small percentage of the variation in risk-adjusted mortality rate can be explained by longitude and latitude. Therefore, other factors beyond geographical location may also play a significant role in determining a hospital's risk-adjusted mortality rate.

Overall, these analyses can provide insight into whether there is any relationship between hospital location and risk-adjusted mortality rate for specific procedures/conditions.
